src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs

```diff
@@ -430,7 +441,1842 @@ unsafe struct S
+                Assert.Equal(module, i2.ContainingModule);
+
+                var c = module.GlobalNamespace.GetTypeMembers("C").Single();
+                var m = c.GetMethod("M");
+
+                var returnType = (FunctionPointerTypeSymbol)m.ReturnType;
+                Assert.Equal(i1, returnType.Signature.ReturnType);
+                Assert.Equal(i2, returnType.Signature.ParameterTypesWithAnnotations[0].Type);
+            }
+        }
+
+        [Fact]
+        public void NoPiaInTypeOf()
+        {
+            var nopiaReference = CreateCompilation(NoPiaInterfaces).EmitToImageReference(embedInteropTypes: true);
+
+            CompileAndVerifyFunctionPointers(@"
+using System;
+unsafe class C
+{
+    public Type M() => typeof(delegate*<I1, I2>);
+}", references: new[] { nopiaReference }, symbolValidator: symbolValidator);
+
+            void symbolValidator(ModuleSymbol module)
+            {
+                Assert.Equal(1, module.ReferencedAssemblies.Length);
+                Assert.NotEqual(nopiaReference.Display, module.ReferencedAssemblies[0].Name);
+
+                var i1 = module.GlobalNamespace.GetTypeMembers("I1").Single();
+                Assert.NotNull(i1);
+                Assert.Equal(module, i1.ContainingModule);
+
+                var i2 = module.GlobalNamespace.GetTypeMembers("I2").Single();
+                Assert.NotNull(i2);
+                Assert.Equal(module, i2.ContainingModule);
+            }
+        }
+
+        [Fact]
+        public void NoPiaInCall()
+        {
+            var nopiaReference = CreateCompilation(NoPiaInterfaces).EmitToImageReference(embedInteropTypes: true);
+
+            var intermediate = CreateCompilation(@"
+using System;
+public unsafe class C
+{
+    public delegate*<I1> M() => throw null;
+}", references: new[] { nopiaReference }, parseOptions: TestOptions.RegularPreview, options: TestOptions.UnsafeReleaseDll).EmitToImageReference();
+
+            CompileAndVerifyFunctionPointers(@"
+unsafe class C2
+{
+    public void M(C c)
+    {
+        _ = c.M()();
+    }
+}", references: new[] { nopiaReference, intermediate }, symbolValidator: symbolValidator);
+
+            void symbolValidator(ModuleSymbol module)
+            {
+                Assert.Equal(2, module.ReferencedAssemblies.Length);
+                Assert.DoesNotContain(nopiaReference.Display, module.ReferencedAssemblies.Select(a => a.Name));
+                Assert.Equal(intermediate.Display, module.ReferencedAssemblies[1].Name);
+
+                var i1 = module.GlobalNamespace.GetTypeMembers("I1").Single();
+                Assert.NotNull(i1);
+                Assert.Equal(module, i1.ContainingModule);
+            }
+        }
+
+        [Fact]
+        public void InternalsVisibleToAccessChecks()
+        {
+            var aRef = CreateCompilation(@"
+using System.Runtime.CompilerServices;
+[assembly: InternalsVisibleTo(""B"")]
+internal class A {}", assemblyName: "A").EmitToImageReference();
```

Comment:
>internal [](start = 0, length = 8)

Does the error go away if the type made public? Could you please add a test for this scenario as well?